synapse/rest/client/room.py

```diff
@@ -204,6 +204,11 @@ async def on_PUT(
         if state_key is not None:
             event_dict["state_key"] = state_key
 
+        # Twisted will have processed the args by now.
+        assert request.args is not None
+        if b"ts" in request.args and requester.app_service:
+            event_dict["origin_server_ts"] = parse_integer(request, "ts", 0)
```

Comment:
What if the provided `origin_server_ts` is a negative number?

Reply:
In the [MSC3316](https://github.com/matrix-org/matrix-doc/pull/3316) this PR is about, the creator mentions that
> Timestamps should already be considered untrusted over federation, and
application services are trusted server components

(see  [here](https://github.com/matrix-org/matrix-doc/blob/555e779f49728b893e37f04ded325457fff07422/proposals/3316-appservice-timestamp-massaging.md?plain=1#L43-L46))

So maybe we can assume that the value is `<0`? Or we could throw an exception? The original lines from which I copied the code also do not handle negative values specifically.

https://github.com/matrix-org/synapse/blob/f0768663eb240f5d95aaad476d890adb4c55e673/synapse/rest/client/room.py#L268-L271